book/src/design/circuit/gadgets/ecc/var-base-scalar-mul.md

```diff
@@ -0,0 +1,258 @@
+
+So, the required check is equivalent to saying that the following "indexed version" of the above algorithm never asserts:
+
+```ignore
+acc := 2
+for i from n-1 down to 0 {
+    p = k_{i+1} ? 1 : −1
+    assert acc ≠ ± p
+    assert (acc + p) ≠ acc    // X
+    acc := (acc + p) + acc
+    assert 0 < acc ≤ (q-1)/2
+}
+if k_0 = 0 {
+    assert acc ≠ 1
+    acc := acc - 1
+}
+```
+
+The maximum value of `acc` is:
+```ignore
+    <--- n 1s --->
+  1011111...111111
+= 1100000...000000 - 1
+```
+= $2^{n+1} + 2^n - 1$
+
+> The assertion labelled X obviously cannot fail because $u \neq 0$. It is possible to see that acc is monotonically increasing except in the last conditional. It reaches its largest value when $k$ is maximal, i.e. $2^{n+1} + 2^n - 1$.
+
+So to entirely avoid exceptional cases, we would need $2^{n+1} + 2^n - 1 < (q-1)/2$. But we can use $n$ larger by $c$ if the last $c$ iterations use [complete addition](./addition.md#Complete-addition).
+
+The first $i$ for which the algorithm using **only** incomplete addition fails is going to be $252$, since $2^{252+1} + 2^{252} - 1 > (q - 1)/2$. We need $n = 254$ to make the wraparound technique above work.
+
+```python
+sage: q = 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001
+sage: 2^253 + 2^252 - 1 < (q-1)//2
+False
+sage: 2^252 + 2^251 - 1 < (q-1)//2
+True
+```
+
+So the last three iterations of the loop ($i = 2..0$) need to use [complete addition](./addition.md#Complete-addition), as does the conditional subtraction at the end. Writing this out using ⸭ for incomplete addition (as we do in the spec), we have:
+
+```ignore
+Acc := [2] T
+for i from 253 down to 3 {
+    P := k_{i+1} ? T : −T
+    Acc := (Acc ⸭ P) ⸭ Acc
+}
+for i from 2 down to 0 {
+    P := k_{i+1} ? T : −T
+    Acc := (Acc + P) + Acc  // complete addition
+}
+return (k_0 = 0) ? (Acc + (-T)) : Acc  // complete addition
+```
+
+## Constraint program for optimized double-and-add (incomplete addition)
+Define a running sum $\mathbf{z_j} = \sum_{i=j}^{n} (\mathbf{k}_{i} \cdot 2^{i-j})$, where $n = 254$ and:
+
+$$
+\begin{aligned}
+    &\mathbf{z}_{n+1} = 0,\\
+    &\mathbf{z}_{n} = \mathbf{k}_{n}, \hspace{2em}\text{(most significant bit)}\\
+	&\mathbf{z}_0 = k.\\
+\end{aligned}
+$$
+
+Initialize $A_{254} = [2] T$
+
+for $i$ from $254$ down to $4$:
+$$
+\begin{aligned}
+    &(\mathbf{k}_i)(\mathbf{k}_i-1) = 0\\
+    &\mathbf{z}_{i} = 2\mathbf{z}_{i+1} + \mathbf{k}_{i}\\
+    & x_{U,i} = x_T\\
+    & y_{U,i} = (2 \mathbf{k}_i - 1) \cdot y_T  \hspace{2em}\text{(conditionally negate)}\\
+    & \lambda_{1,i} \cdot (x_{A,i} - x_{U,i}) = y_{A,i} - y_{U,i}\\
+    & \lambda_{1,i}^2 = x_{R,i} + x_{A,i} + x_{U,i}\\
```

Comment:
Everywhere else on this page we use P instead of U:
```suggestion
    & x_{P,i} = x_T\\
    & y_{P,i} = (2 \mathbf{k}_i - 1) \cdot y_T  \hspace{2em}\text{(conditionally negate)}\\
    & \lambda_{1,i} \cdot (x_{A,i} - x_{P,i}) = y_{A,i} - y_{P,i}\\
    & \lambda_{1,i}^2 = x_{R,i} + x_{A,i} + x_{P,i}\\
```